src/bert/utils.py

```diff
@@ -0,0 +1,102 @@
+import logging as log
+
+from typing import Dict
+
+import torch
+import torch.nn as nn
+
+# huggingface implementation of BERT
+import pytorch_pretrained_bert
+
+class BertEmbedderModule(nn.Module):
+    """ Wrapper for BERT module to fit into jiant APIs. """
+
+    def __init__(self, args, cache_dir=None):
+        super(BertEmbedderModule, self).__init__()
+
+        self.model = \
+            pytorch_pretrained_bert.BertModel.from_pretrained(
+                args.bert_model_name,
+                cache_dir=cache_dir)
+        self.embeddings_mode = args.bert_embeddings_mode
+        # BERT supports up to 512 tokens; see section 3.2 of https://arxiv.org/pdf/1810.04805.pdf
+        assert args.max_seq_len <= 512
+        self.seq_len = args.max_seq_len
+
+        # Set trainability of this module.
+        for param in self.model.parameters():
+            param.requires_grad = bool(args.bert_fine_tune)
+
+    def forward(self, sent: Dict[str, torch.LongTensor],
+                unused_task_name: str="") -> torch.FloatTensor:
+        """ Run BERT to get hidden states.
+
+        Args:
+            sent: batch dictionary
+
+        Returns:
+            h: [batch_size, seq_len, d_emb]
+        """
+        assert "bert_wpm_pretokenized" in sent
+        # <int32> [batch_size, var_seq_len]
+        var_ids = sent["bert_wpm_pretokenized"]
+
+        # Model has fixed, learned positional component, so we must pass a
+        # block of exactly seq_len length.
+        # ids is <int32> [batch_size, seq_len]
+        ids = torch.zeros(var_ids.size()[0], self.seq_len, dtype=var_ids.dtype,
+                          device=var_ids.device)
+        fill_len = min(var_ids.size()[1], self.seq_len)
+        ids[:,:fill_len] = var_ids[:,:fill_len]
+
+        mask = (ids == 0)
+        # "Correct" ids to account for different indexing between BERT and
+        # AllenNLP.
+        # The AllenNLP indexer adds a '@@UNKNOWN@@' token to the
+        # beginning of the vocabulary, *and* treats that as index 1 (index 0 is
+        # reserved for padding).
+        FILL_ID = 0  # [PAD] for BERT models.
+        ids[ids == 0] = FILL_ID + 2
+        ids -= 2
+
+        if self.embeddings_mode != "none":
+            # Extract lexical embeddings; see
+            # https://github.com/huggingface/pytorch-pretrained-BERT/blob/master/pytorch_pretrained_bert/modeling.py#L186
+            h_lex = self.model.embeddings.word_embeddings(ids)
+            h_lex = self.model.embeddings.LayerNorm(h_lex)
+            # following our use of the OpenAI model, don't use dropout for
+            # probing.
+            #  h_lex = self.model.embeddings.dropout(embeddings)
```

Comment:
Maybe make it an option whether or not the model is being used in a probing context? I think we'll (read: I will want to) use this in a non-probing context.
Also, I forget is the convention is to apply dropout at the end of a module or at the beginning of the next use (I remember this caused an issue during the workshop).

Reply:
FWIW, this is basically a shortcut to treating the BERT WPM like GloVe vectors. Erring on the side of no dropout here, since there are a bunch of places we can add it later if these are the only embeds?

I'd rather not add an additional "is_probing" param since that introduces a lot of cruft.

Reply:
:+1: